Implement glfw joystick API

satoshinm opened the pull request:
Adds support for glfw3's joystick functions: http://www.glfw.org/docs/latest/input_guide.html#joystick

- [x] glfwJoystickPresent
- [x] glfwGetJoystickAxes
- [x] glfwGetJoystickButtons
- [x] glfwGetJoystickName
- [x] glfwSetJoystickCallback

Includes a test based on the sdl joystick test:

```sh
python tests/runner.py browser.test_glfw_joystick
```

Also manually tested with a real joystick and a more complex real-world application which had native glfw joystick support (https://github.com/satoshinm/NetCraft/pull/80)

---

- [x] Add _free()s on gamepaddisconnect, .id .buttons
- [x] Add _free()s on gamepaddisconnect, .axes
- [x] Incorporate most review feedback
- [x] Browser test: `python tests/runner.py browser.test_glfw_joystick`
- [x] Interactive test: `python tests/runner.py interactive.test_glfw_joystick`
- [x] Test versus emscripten gamepad API, and native glfw API
- [x] Poll once per frame, see https://github.com/kripken/emscripten/pull/4292

satoshinm commented:
This is functional but is it ok to refresh the joystick status the calls to glfwGetJoystickAxes and glfwGetJoystickButtons? Wasn't sure if the joystick state should be read on each frame instead. Also do I need to free the memory allocated for the axes and buttons arrays (from `allocate`, couldn't find out how to free it)

kripken commented:
I'm not sure about the refreshing, but looks like it might be fast enough. I guess if you don't see it show up in profiling, it's probably fine. Or was the concern something other than speed?

For the allocations, `ALLOC_NORMAL` uses malloc, so you can do `_free(..)`.

juj commented on src/library_glfw.js:
```diff
@@ -57,6 +56,9 @@ var LibraryGLFW = {
       this.keys = new Array();
+      this.joyLast = 0; // GLFW_JOYSTICK_1
+      this.joys = {};
+      this.index2joy = {};
```
Can you document here what is the purpose of these three variables. The name `index2joy` is a bit ad hoc, it would be good to have some details here.

juj commented on src/library_glfw.js:
```diff
@@ -382,6 +385,38 @@ var LibraryGLFW = {
     },
 
+    onGamepadConnected: function(event) {
+      if (!GLFW.active.joystickFunc) return;
```
I wonder if GLFW should be tracking gamepads already before the user code has registered a joystick callback function. For example, it can take a while until the user code runs, and a gamepad connected event may have fired already before. Or perhaps the tightest way would be to enumerate all already known joysticks when a callback is registered or GLW initialized to initially populate the already known joysticks. I'm thinking about the scenario where user is clicking on gamepad buttons way before the page has fully loaded up.

juj commented on src/library_glfw.js:
```diff
@@ -382,6 +385,38 @@ var LibraryGLFW = {
+
+      var joy = GLFW.active.index2joy[event.gamepad.index];
+
+      Module['dynCall_vii'](GLFW.active.joystickFunc, joy, 0x00040002); // GLFW_DISCONNECTED
```
What if `joy` is `undefined` here? I suppose the callback should not be called then if such an odd situation arises?

juj commented on src/library_glfw.js:
```diff
@@ -382,6 +385,38 @@ var LibraryGLFW = {
+
+      Module['dynCall_vii'](GLFW.active.joystickFunc, joy, 0x00040002); // GLFW_DISCONNECTED
+
+      // TODO: free memory .axes, .buttons
```
This can do `_free(GLFW.active.joys[joy].id);` and same for `buttons` and `axes`. (`ALLOC_NORMAL` is equivalent to a `_malloc()`)

juj commented on src/library_glfw.js:
```diff
@@ -382,6 +385,38 @@ var LibraryGLFW = {
+    },
+
+    onGamepadDisconnected: function(event) {
+      if (!GLFW.active.joystickFunc) return;
```
Here also if user code has already detached a user callback function, the code below should still run to deinitialize the data for that joystick?

juj commented on src/library_glfw.js:
```diff
@@ -641,6 +676,31 @@ var LibraryGLFW = {
+      var j = GLFW.active.joys[joy];
+      if (!j) return;
+
+      var gamepad = navigator.getGamepads()[j.index];
```
Reading the documentation at http://www.glfw.org/docs/latest/group__input.html#gace54cd930dcd502e118fe4021384ce1b and http://www.glfw.org/docs/latest/input_guide.html#joystick_button, it looks like a suitable usage from user code might be to do a 

    for(int i = 0; i < numJoysticks; ++i) {
        int count;
        const unsigned char* axes = glfwGetJoystickButtons(GLFW_JOYSTICK_1+i, &count);
        for(int j = 0; j < count; ++j) {
            // process axes[j];
        }
    }

In that case, there is a possible subtle bug due to an unspecced feature of the Gamepad API regarding `navigator.getGamepads()`, see https://github.com/w3c/gamepad/issues/22 and https://github.com/w3c/gamepad/issues/8. The issue is that `navigator.getGamepads()` polls new state for all gamepads, but the above line only reads data from one of them. So if there are e.g. 3 joysticks, the above loop would poll the state of each of the three joysticks three times, and process only one of them at each loop iteration. This might potentially miss out button presses, but whether that happens in practice is somewhat theoretical. Still good to keep in mind though.

juj replied:
See https://github.com/kripken/emscripten/pull/4292 for a similar change that was done for the HTML5 API counterpart, and also https://github.com/kripken/emscripten/pull/4921 which might be of interest.

juj commented on src/library_glfw.js:
```diff
@@ -1354,15 +1418,43 @@ var LibraryGLFW = {
+  },
+
+  glfwGetJoystickName: function(joy) {
+    return GLFW.active.joys[joy].id;
```
If joystick is disconnected right before this function is called, `GLFW.active.joys[joy]` could be deleted above and be undefined, making this function throw an exception on `undefined.id` access; so check for `if (GLFW.active.joys[joy])` here before proceeding?

juj commented on tests/test_browser.py:
```diff
@@ -1076,6 +1076,53 @@ def test_sdl_joystick_2(self):
     Popen([PYTHON, EMCC, os.path.join(self.get_dir(), 'sdl_joystick.c'), '-O2', '--minify', '0', '-o', 'page.html', '--pre-js', 'pre.js', '-lSDL', '-lGL']).communicate()
     self.run_browser('page.html', '', '/report_result?2')
 
+  def test_glfw_joystick(self):
```
Great to have a mock input injecting test! It would be nice to add the same test to the `interactive` suite as well, but without the API mock implementation part, so that could be tested with a live gamepad.

satoshinm replied:
Added a simple interactive test (run with `python tests/runner.py interactive.test_glfw_joystick`), no graphics it only logs text of the gamepad state changes but useful for testing (already found & fixed the axes memory allocation error thanks to it)

satoshinm commented:
Incorporated most (but not all) of the review feedback. Next major change needed is to poll the gamepad state once per frame like in https://github.com/kripken/emscripten/pull/4292. This should also solve the issues I'm seeing on Chrome where gamepadconnected is fired on disconnect. Wondering if it is worth preserving this API nuance of glfw:

```c
      // HTML5 Gamepad API reuses indexes
      // glfw joystick API spreads them out
      // http://www.glfw.org/docs/latest/input_guide.html#joystick
      // "Once a joystick is detected, it keeps its assigned index until it is disconnected,
      // so as joysticks are connected and disconnected, they will become spread out."
```

Also I added _free() but not to `_free(GLFW.active.joys[joy].axes);`, crashes in abort() on corrupted memory, need to investigate more what went wrong there...

satoshinm commented:
Changed to sample once per frame, tracking joystick connect/disconnect on refresh, and fixed the axes memory freeing. I think that's everything, anything else?

satoshinm commented:
@kripken @juj can this be merged or anything I need to change first?

kripken commented:
This looks good to me, but let's wait a bit more for @juj.

kripken commented:
Looks like @juj has no immediate concerns here, so let's merge this, any further stuff can be done in followup work.
